### DISCUSSION TITLE
How to find information and read the documentation?

### DISCUSSION THREAD

**Discussion by someparsa** [IfcOpenShell/IfcOpenShell#2715] [General]
I have a question about the documentation. How should I read it?

As an instance, in this page ([link](https://blenderbim.org/docs-python/ifcopenshell-python/code_examples.html#create-a-simple-model-from-scratch)), we have an example on producing a model from scratch.

```
ifcopenshell.api.run("geometry.add_wall_representation", model,
    context=body, length=5, height=3, thickness=0.2)
```

I searched the documentation so much. I find no clue where length, height, and thickness come from. I know they are characteristics of a wall. But they are not introduced as options or requirements of the representation in the documentation. This is the only info we find in the site ([link](https://blenderbim.org/docs-python/autoapi/ifcopenshell/api/geometry/add_wall_representation/index.html)):

`class ifcopenshell.api.geometry.add_wall_representation.Usecase(file, **settings)`

Length, height, and thickness should be `**settings` which are never made clear in the documentation, what they are.

Now, if I want to make a beam or column, what characteristics should I introduce?

I assume these are properties or parameters which are obvious in other language like IFC C# or C++ and the documentation is produced in a way, the user should understand those basics from other references. So the API is written in a higher level, neglecting this obvious information.

Could you please advise me how I can find these functions?

**Comment by Moult** [comment]
The geometry module is the only remaining module that is not documented, which is why you can't find what you're looking for. It'll probably be documented for the next release. For now, you can read the code: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/geometry/add_wall_representation.py#L27-L37

**Reply by someparsa** [reply]
That is awesome. Thanks for sharing. Looking forward to reading the updates. Any help I can provide, please let me know.

**Reply by aothms** [reply]
And another quick comment:

> Now, if I want to make a beam or column, what characteristics should I introduce?

I think would be https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/geometry/add_profile_representation.py

**Reply by someparsa** [reply]
Thank you aothms. It is great. The link provides an example of defining a cross-section. What I was looking for, is defining a line element (let's say a beam or column) by defining its two end nodes (by coordinates). I think I should wait for the documentation or dig in the back-end codes as Moult mentioned earlier.

**Reply by aothms** [reply]
That would be https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/geometry/add_axis_representation.py Typically elements in IFC do have a Body representation though, which is as you said basically the cross section swept (IfcExtrudedAreaSolid) along the axis (2d / 3d curve). An element in IFC can (and is recommended to have) both of these two representations.

**Reply by someparsa** [reply]
That answers my concern then, thanks for sharing. And the command is always like, we define geometry, and then we assign the cross-section. That is reasonable and similar to the Wall example we have in the documentation.

BTW, what I see is that all the scripts are defined in the same way.
`ifcopenshell.api.run(API text in "", model, context, options])`

This has two aspects. It is easy to define and helps to do things faster for the long-term user. But at the first sight it becomes tough for native Python developers. Because it is different from most of the Pythonic libraries, we cannot hover over the code to get hints on the commands, the documentation is different, etc.

Anyway, I know that this is a very big project and projects like this mostly have their own technical culture to do the job.

**Reply by aothms** [reply]
See https://github.com/IfcOpenShell/IfcOpenShell/issues/2693 All help is welcome :)

**Reply by someparsa** [reply]
Hi Thomas, reading the python references you referred to above, I believe I need to define the placement of an entity I make. For instance, I need to place the starting node of a beam at Cartesian coordinate of (x0, y0, z0). I can see there are edit_object_placement keywords between the code lines, but I have no idea how I may use them.

**Reply by Moult** [reply]
By the way, although the geometry module is not documented, you will find it in use in various other code examples in other modules, such as the material module (for parametric material profiles). Here's one example of creating a beam: https://blenderbim.org/docs-python/autoapi/ifcopenshell/api/material/edit_profile_usage/index.html

There are also three examples of setting the object placement here: https://blenderbim.org/docs-python/autoapi/ifcopenshell/api/void/add_filling/index.html basically you give it a transformation matrix. No matrix means the identity matrix (at the origin, no rotation).

**Reply by someparsa** [reply]
Thanks Dion. BTW, if I know how I should read your functions and definitions, I will gradually document the remaining for you in my free time. Just need to get on track and will do the rest.

**Reply by aothms** [reply]
Placements in IFC are a bit specific. There is a hierarchy of placements that follows the subdivision structure of the project (Site > Building > Storey > Wall > Opening > Window for example). That's IfcLocalPlacement. The relative positioning is made using an "axis 2 placement" (inherited from step) which is a point and 2 direction vectors (y is implied as x cross z). I think in quite a few cases the API is built around 4x4 numpy matrices to encode the placements/positioning. Other than that I don't have any specific insights, but if you have specific questions we can answer.

**Reply by someparsa** [reply]
Thanks, Thomas. I think I should read the blenderbim api which Dion referred to in his comment for a few days too.
